pkg/tsdb/cloudwatch/live.go

```diff
@@ -0,0 +1,254 @@
+	"golang.org/x/sync/errgroup"
+)
+
+type LogQueryRunnerSupplier struct {
+}
+
+type logQueryRunner struct {
+	channelName string
+	publish     models.ChannelPublisher
+	running     map[string]bool
+}
+
+const (
+	maxAttempts   = 8
+	minRetryDelay = 500 * time.Millisecond
+	maxRetryDelay = 30 * time.Second
+)
+
+var (
+	channelMu        = sync.Mutex{}
+	responseChannels = map[string]chan *tsdb.Response{}
+)
+
+func addResponseChannel(name string, channel chan *tsdb.Response) error {
+	channelMu.Lock()
+	defer channelMu.Unlock()
+
+	if _, ok := responseChannels[name]; ok {
+		return fmt.Errorf("channel with name '%s' already exists", name)
+	}
+
+	responseChannels[name] = channel
+	return nil
+}
+
+func getResponseChannel(name string) (chan *tsdb.Response, error) {
+	channelMu.Lock()
+	defer channelMu.Unlock()
+
+	if responseChannel, ok := responseChannels[name]; ok {
+		return responseChannel, nil
+	}
+
+	return nil, fmt.Errorf("channel with name '%s' not found", name)
+}
+
+// GetHandlerForPath gets called on init.
+func (supplier *LogQueryRunnerSupplier) GetHandlerForPath(path string, publisher models.ChannelPublisher) (models.ChannelHandler, error) {
+	return &logQueryRunner{
+		channelName: path,
+		publish:     publisher,
+		running:     make(map[string]bool),
+	}, nil
+}
+
+// GetChannelOptions gets channel options.
+// It's called fast and often.
+func (g *logQueryRunner) GetChannelOptions(id string) centrifuge.ChannelOptions {
+	return centrifuge.ChannelOptions{}
+}
+
+// OnSubscribe publishes results from the corresponding CloudWatch Logs query to the provided channel
+func (g *logQueryRunner) OnSubscribe(c *centrifuge.Client, e centrifuge.SubscribeEvent) error {
+	if _, ok := g.running[e.Channel]; !ok {
+		g.running[e.Channel] = true
+		go func() {
+			if err := g.publishResults(e.Channel); err != nil {
+				plog.Error(err.Error())
+			}
+		}()
+	}
+
+	return nil
+}
+
+// OnPublish is called when an event is received from the websocket.
+func (g *logQueryRunner) OnPublish(c *centrifuge.Client, e centrifuge.PublishEvent) ([]byte, error) {
+	return nil, fmt.Errorf("can not publish")
+}
+
+func (g *logQueryRunner) publishResults(channelName string) error {
+	defer func() {
+		delete(responseChannels, channelName)
+		delete(g.running, channelName)
+	}()
+
+	responseChannel, err := getResponseChannel(channelName)
+	if err != nil {
+		return err
+	}
+
+	for response := range responseChannel {
+		responseBytes, err := json.Marshal(response)
+		if err != nil {
+			return err
+		}
+
+		if err := g.publish(channelName, responseBytes); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *cloudWatchExecutor) executeLiveLogQuery(ctx context.Context, queryContext *tsdb.TsdbQuery) (*tsdb.Response, error) {
+	responseChannelName := uuid.Must(uuid.NewV4()).String()
+	responseChannel := make(chan *tsdb.Response)
+	if err := addResponseChannel("plugin/cloudwatch/"+responseChannelName, responseChannel); err != nil {
+		return nil, err
+	}
```

Comment:
This will create a new channel for each request -- this may be fine, but also means that concurrent users all make individual queries to the data source.  If the query is "show me the latest logs", ideally we can make a *single* request to cloudwatch, and share the results with everyone making the same query.  Given that queries come from dashboards, it is very likely to have everyone executing the same query!

If there this is an easy change, base64(part of the query + salt)?, i think we should try soon -- if not multiple queries is fine till we figure out a reasonable strategy.

Reply:
In theory I like the idea of being able to share results among multiple users, but since query results are informed by start and end time ranges, in practice I imagine it's unlikely enough that many users would have the exact time ranges & queries that I'm reluctant to optimize for that use case.

Reply:
:+1: -- it really depends on what a normal query is and how long they are expected to run. 

I think the current state is good to merge (likely after some changes from @aknuds1), but we should revist some channel options in 7.3+:
* private channels -- with the current query model, we are creating a new channel for each request, but it is only ever sent to one person.  I think there is a method that we only publish results back to an explicit client Id
* consider reuse -- for splunk logs, we will certainly want to reuse the query results for all connected clients (queries are very expensive), I assume we can more easily improve the pattern here before adding the complication for how to real plugins